ethereumj-core/src/main/java/org/ethereum/sync/SyncQueueImpl.java

```diff
@@ -360,6 +367,80 @@ private HeadersRequestImpl getNextReq(long startFrom, int maxCount) {
         return null;
     }
 
+    @Override
+    public ValidatedHeaders addHeadersAndValidate(Collection<BlockHeaderWrapper> headers) {
+        for (BlockHeaderWrapper header : headers) {
+            addHeader(header);
+        }
+
+        List<HeaderElement> longestChain = getLongestChain();
+        ValidatedHeaders result = validateChain(longestChain);
+        if (result.isValid()) {
+            trimChainImpl(longestChain);
+        } else {
+            // erase chain starting from first invalid header
+            eraseChain(longestChain, result.getHeaders().get(0).getNumber());
+        }
+
+        return result;
+    }
+
+    /**
+     * Runs parent header validation and returns after first occurrence of invalid header
+     */
+    ValidatedHeaders validateChain(List<HeaderElement> chain) {
+        if (chain.size() <= MAX_CHAIN_LEN || parentHeaderValidator == null)
```

Comment:
Could you, please, add comment why we could pass <= MAX_CHAIN_LEN

Reply:
I could add a comment. But this is a result of a trade off between bad design and changes in legacy code. And I think I've found the way out.